agent/consul/grpc_resolver.go

```diff
@@ -0,0 +1,226 @@
+	resolvers  map[string]*ServerResolver
+	shutdownCh <-chan struct{}
+	lock       sync.Mutex
+}
+
+func NewServerResolverBuilder(scheme, datacenter string, shutdownCh <-chan struct{}) *ServerResolverBuilder {
+	return &ServerResolverBuilder{
+		scheme:     scheme,
+		datacenter: datacenter,
+		servers:    make(map[string]*metadata.Server),
+		resolvers:  make(map[string]*ServerResolver),
+	}
+}
+
+// periodicServerRebalance periodically reshuffles the order of server addresses
+// within the resolvers to ensure the load is balanced across servers.
+func (s *ServerResolverBuilder) periodicServerRebalance(serf *serf.Serf) {
+	// Compute the rebalance timer based on the number of local servers and nodes.
+	rebalanceDuration := router.ComputeRebalanceTimer(s.serversInDC(s.datacenter), serf.NumNodes())
+	timer := time.NewTimer(rebalanceDuration)
+
+	for {
+		select {
+		case <-timer.C:
+			s.rebalanceResolvers()
+
+			// Re-compute the wait duration.
+			newTimerDuration := router.ComputeRebalanceTimer(s.serversInDC(s.datacenter), serf.NumNodes())
+			timer.Reset(newTimerDuration)
+		case <-s.shutdownCh:
+			timer.Stop()
+			return
+		}
+	}
+}
+
+// rebalanceResolvers shuffles the server list for resolvers in all datacenters.
+func (s *ServerResolverBuilder) rebalanceResolvers() {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for _, resolver := range s.resolvers {
+		// Shuffle the list of addresses using the last list given to the resolver.
+		resolver.addrLock.Lock()
+		addrs := resolver.lastAddrs
+		rand.Shuffle(len(addrs), func(i, j int) {
+			addrs[i], addrs[j] = addrs[j], addrs[i]
+		})
+		resolver.addrLock.Unlock()
+
+		// Pass the shuffled list to the resolver.
+		resolver.updateAddrs(addrs)
+	}
+}
+
+// serversInDC returns the number of servers in the given datacenter.
+func (s *ServerResolverBuilder) serversInDC(dc string) int {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	var serverCount int
+	for _, server := range s.servers {
+		if server.Datacenter == dc {
+			serverCount++
+		}
+	}
+
+	return serverCount
+}
+
+// Servers returns metadata for all currently known servers. This is used
+// by grpc.ClientConn through our custom dialer.
+func (s *ServerResolverBuilder) Servers() []*metadata.Server {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	servers := make([]*metadata.Server, 0, len(s.servers))
+	for _, server := range s.servers {
+		servers = append(servers, server)
+	}
+	return servers
+}
+
+// Build returns a new ServerResolver for the given ClientConn. The resolver
+// will keep the ClientConn's state updated based on updates from Serf.
+func (s *ServerResolverBuilder) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOption) (resolver.Resolver, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// If there's already a resolver for this datacenter, return it.
+	datacenter := strings.TrimPrefix(target.Endpoint, "server.")
+	if resolver, ok := s.resolvers[datacenter]; ok {
+		return resolver, nil
+	}
```

Comment:
One issue found while trying to reproduce benchmarks from earlier is that our GRPCConn handling is a bit confusing because of this.

The issue comes from the caching of ClientConns by DC in our custom resolver here (which is global to a process).

If you have multiple `GRPCClient` objects in a single process, each one will have a separate set of `ClientConns` even for the same DC. But our resolver builder will return the same Resolver for each of them and that resolver will be wrapping the first `ClientConn` from the first Client. That means that the resolver will only ever deliver server endpoints to the first `ClientConn` that connects and all others will hang forever on an RPC call waiting for resolution.

It's not directly a bug here because in normal use we only ever have a single `GRPCClient` instance in the Consul agent and so the assumptions hold. But it feels confusing and brittle.

I wonder if we should just fix it - I'm not sure the caching here is really that necessary. Even if we keep it we could instead key it by datacenter AND `cc` pointer address or something so that ever `ClientConn` that attempts to resolve gets delivered updates even if there are multiple for the same DC.

I'm in two minds. On the one hand this feels confusing and like it might trip us up at some point if we introduce a second `GRPCClient` into a new component for some reason or run multiple in parallel tests for example.

It is preventing me updating the bench mark tool though since we can't simulate multiple clients (separate sets of http2 conns) in a single process.

I think the best thing would be to change this so we key the map by the `ClientConn` instance instead.